File: year_2021/solutions/julia/day_05.jl
include(joinpath(@__DIR__(), "utils", "files.jl"))


struct Point
    x::Int
    y::Int
end


struct Segment
    start_::Point
    end_::Point
end


function count_segment_overlaps(segments::Array{Segment}, use_diagonals::Bool)::Int
    visits = Dict{Point, Int}()
    for segment in segments
        if !use_diagonals & !is_horizontal_or_vertical(segment)
            continue
        end
        for point in get_points(segment)
            visits[point] = point in keys(visits) ? visits[point] + 1 : 1
        end
    end
    return length([p for p in keys(visits) if visits[p] > 1])
end


get_points(segment::Segment)::Array{Point} = [
    Point(x, y) 
    for (x, y) in ziplongest(segment.start_.x:x_growth_step(segment):segment.end_.x, segment.start_.y:y_growth_step(segment):segment.end_.y)
]

is_horizontal_or_vertical(segment::Segment)::Bool = (segment.start_.x == segment.end_.x) | (segment.start_.y == segment.end_.y)

x_growth_step(segment::Segment)::Int = segment.start_.x <= segment.end_.x ? 1 : -1

y_growth_step(segment::Segment)::Int = segment.start_.y <= segment.end_.y ? 1 : -1


function ziplongest(range_x::StepRange, range_y::StepRange)
    if (length(range_x) == 1) | (length(range_y) == 1)
        return [(x, y) for x in range_x for y in range_y]
    elseif (length(range_x) == length(range_y))
        return zip(range_x, range_y)
    else
        throw(ErrorException("The ranges $range_x and $range_y have different sizes and cannot be zipped."))
    end
end


read_segments(input_file_path::String)::Array{Segment} = [
    Segment(map(p -> Point(parse.(Int, split(p, ","))...), split(line, " -> "))...) 
    for line in readlines(input_file_path)
]


function main()
    input_file_path::String = joinpath(INPUTS_FOLDER, splitext(basename(@__FILE__))[1], "input.txt")
    segments::Array{Segment} = read_segments(input_file_path)

    # Part 1
    part_1_result::Int = count_segment_overlaps(segments, false)
    @assert part_1_result == 5774
    println("Part 1 result : ", part_1_result)

    # Part 2
    part_2_result::Int = count_segment_overlaps(segments, true)
    @assert part_2_result == 18423
    println("Part 2 result : ", part_2_result)
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: year_2021/solutions/python/day_05.py
from dataclasses import dataclass
from pathlib import Path
from typing import List

from year_2021.solutions.python.utils.files import INPUTS_FOLDER


def count_segment_overlaps(segments: List['Segment'], use_diagonals: bool) -> int:
    visits_counts = {}
    for segment in segments:
        if not use_diagonals and segment.direction == 'd':
            continue
        for point in segment.get_points():
            if point in visits_counts:
                visits_counts[point] += 1
            else:
                visits_counts[point] = 1
    return len([point for point, visits in visits_counts.items() if visits > 1])


@dataclass(frozen=True)
class Point:
    x: int
    y: int


class Segment:

    def __init__(self, start: Point, end: Point):
        self.start = start
        self.end = end
        if self.start.y == self.end.y:
            self.direction = 'h'
        elif self.start.x == self.end.x:
            self.direction = 'v'
        elif abs(self.end.x - self.start.x) == abs(self.end.y - self.start.y):
            self.direction = 'd'
        else:
            raise ValueError('Invalid segment. Each segment should be either horizontal, vertical or diagonal.')

    def get_points(self) -> List[Point]:
        match self.direction, self.is_increasing('x'), self.is_increasing('y'):
            case 'h', True, _:
                return [Point(x, self.start.y) for x in range(self.start.x, self.end.x + 1)]
            case 'h', False, _:
                return [Point(x, self.start.y) for x in range(self.start.x, self.end.x - 1, -1)]
            case 'v', _, True:
                return [Point(self.start.x, y) for y in range(self.start.y, self.end.y + 1)]
            case 'v', _, False:
                return [Point(self.start.x, y) for y in range(self.start.y, self.end.y - 1, -1)]
            case 'd', True, True:
                return [Point(x, y)
                        for x, y in zip(range(self.start.x, self.end.x + 1), range(self.start.y, self.end.y + 1))]
            case 'd', True, False:
                return [Point(x, y)
                        for x, y in zip(range(self.start.x, self.end.x + 1), range(self.start.y, self.end.y - 1, -1))]
            case 'd', False, True:
                return [Point(x, y)
                        for x, y in zip(range(self.start.x, self.end.x - 1, -1), range(self.start.y, self.end.y + 1))]
            case 'd', False, False:
                return [Point(x, y)
                        for x, y in
                        zip(range(self.start.x, self.end.x - 1, -1), range(self.start.y, self.end.y - 1, -1))]
            case _, _, _:
                raise ValueError("Invalid segment.")

    def is_increasing(self, coordinate: str):
        match coordinate:
            case 'x':
                return self.start.x <= self.end.x
            case 'y':
                return self.start.y <= self.end.y


def read_segments(file_path: Path) -> List[Segment]:
    segments = []
    with open(file_path, 'r') as input_file:
        for line in input_file.read().split('\n'):
            start_point, end_point = line.split(' -> ')
            start_x, start_y = start_point.split(',')
            end_x, end_y = end_point.split(',')
            segments.append(
                Segment(
                    Point(int(start_x), int(start_y)),
                    Point(int(end_x), int(end_y))
                )
            )
    return segments


if __name__ == '__main__':
    input_file_path = INPUTS_FOLDER / Path(__file__).with_suffix('').name / 'input.txt'
    all_segments: List[Segment] = read_segments(file_path=input_file_path)

    # Part 1
    part_1_result: int = count_segment_overlaps(all_segments, use_diagonals=False)
    assert part_1_result == 5774
    print('Part 1 result :', part_1_result)

    # Part 2
    part_2_result: int = count_segment_overlaps(all_segments, use_diagonals=True)
    assert part_2_result == 18423
    print('Part 2 result :', part_2_result)
